src/main/java/edu/kit/compiler/assembly/JumpInversion.java

```diff
@@ -0,0 +1,171 @@
+package edu.kit.compiler.assembly;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import edu.kit.compiler.assembly.AssemblyOptimizer.AssemblyOptimization;
+import edu.kit.compiler.io.BufferedLookaheadIterator;
+import edu.kit.compiler.io.LookaheadIterator;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * An optimization to improve control flow in assembly. A conditional jump
+ * to Label A, followed by an unconditional jump to label B is replaced with
+ * a single conditional jump to label B if fallthrough is possible.
+ */
+public final class JumpInversion implements AssemblyOptimization {
+
+    @Override
+    public Iterator<String> optimize(Iterator<String> instructions) {
+        return new Optimization(new BufferedLookaheadIterator<>(instructions));
+    }
+
+    @RequiredArgsConstructor
+    private static final class Optimization implements Iterator<String> {
+
+        private final LookaheadIterator<String> input;
+
+        @Override
+        public boolean hasNext() {
+            return input.has();
+        }
+
+        @Override
+        public String next() {
+            if (input.has(2)) {
+                return parseConditional();
+            } else {
+                return input.getNext();
+            }
+        }
+
+        private String parseConditional() {
+            assert input.has(0);
+
+            var instruction = input.get(0);
+            var space = instruction.indexOf(' ');
+
+            if (space != -1) {
+                return Jump.get(instruction.substring(0, space))
+                        .map(jump -> {
+                            var trueLabel = instruction.substring(space).stripLeading();
+                            return parseUnconditional(jump, trueLabel);
+                        })
+                        .orElseGet(input::getNext);
+            } else {
+                return input.getNext();
+            }
+        }
+
+        private String parseUnconditional(Jump trueJump, String trueLabel) {
+            assert input.has(1);
+
+            var instruction = input.get(1);
+
+            if (instruction.startsWith("jmp ")) {
+                var falseLabel = instruction.substring(3).stripLeading();
+                return parseLabel(trueJump, trueLabel, falseLabel);
+            } else {
+                return input.getNext();
+            }
+        }
+
+        private String parseLabel(Jump trueJump, String trueLabel, String falseLabel) {
+            assert input.has(2);
+
+            var instruction = input.get(2);
+            if (instruction.startsWith(trueLabel) && instruction.endsWith(":")) {
+                input.next(2);
+                return trueJump.getInverse().getInstruction() + " " + falseLabel;
```

Comment:
This merges the instruction and the label into one line, right? This is probably not wrong, considering the way we handle the assembly, but it is rather ugly...